django_emarsys/managers.py

```diff
@@ -2,81 +2,94 @@
+
+        event = self.create(
+            event_name=event_name,
+            recipient_email=recipient_email,
+            user=user,
+            source=source)
+
+        if data is None:
+            data = {}
 
         try:
-            emarsys_events = api.get_events()
-            for local_event in self.all():
-                emarsys_event_id = emarsys_events.get(local_event.name)
-                if emarsys_event_id:
-                    if not local_event.emarsys_id:
-                        num_added_ids += 1
-                    elif local_event.emarsys_id != emarsys_event_id:
-                        num_updated_ids += 1
-
-                    local_event.emarsys_id = emarsys_event_id
-                    del emarsys_events[local_event.name]
-                else:
-                    if local_event.emarsys_id:
-                        num_deleted_ids += 1
-                    local_event.emarsys_id = None
-
-                local_event.save()
-
-            for emarsys_event_name, emarsys_event_id in emarsys_events.items():
-                self.create(emarsys_id=emarsys_event_id, name=emarsys_event_name)
-                num_new_events += 1
-        except EmarsysError as e:
-            sys.stderr.write('Emarsys error: {}\n'.format(e))
+            context = {'global': {
+                key: conditional_escape(value)
+                for key, value
+                in self.get_placeholder_data(event_name,
+                                             **dict(data, user=user)).items()}}
 
-        return num_new_events, num_added_ids, num_updated_ids, num_deleted_ids
+            event.set_context(context)
 
-    def trigger(self, event_name, recipient_email, user=None, data=None,
-                create_user_if_needed=True, async=False, get_create_data=None):
-        event = self.get(name=event_name)
-        event_instance = event.trigger(
-            recipient_email, user, data, async=async)
+            if event_name not in settings.EMARSYS_EVENTS:
+                raise UnknownEventNameError(event_name)
+
+            event.set_data(data)
+        except Exception as e:
+            event.handle_error(e)
+            return event
 
-        if (event_instance.state == 'error'
-                and event_instance.result_code == '2008'
-                and create_user_if_needed):
+        if getattr(settings, 'EMARSYS_RECIPIENT_WHITELIST', None) is not None:
+            if recipient_email not in settings.EMARSYS_RECIPIENT_WHITELIST:
+                event.handle_error("User not on whitelist: {}"
+                                   .format(recipient_email))
+                return event
+
+        from django_emarsys.models import EmarsysEventId
+        try:
+            event_id = EmarsysEventId.objects.get(event_name=event_name)
+        except EmarsysEventId.DoesNotExist:
+            log.warning("Emarsys-ID for event '{event}' unknown, "
+                        "syncing...".format(event=event_name))
```

Comment:
Warning might be too strong here. If syncing gets the event, that's a normal and expected case, i.e. info or even debug. If the event isn't there even after syncing, you get a stronger log later (I believe? you should.).